File: dwarf/targets/cardano-node/src/DecodeChainSync.hs
{-# LANGUAGE OverloadedStrings #-}
-- | cardano-node-mini-protocol-decode-chainsync shim.
-- Decodes one ChainSync mini-protocol wire-format message.
module Main (main) where

import           Control.Monad                    (replicateM_)
import qualified Codec.CBOR.Decoding              as CBOR
import qualified Codec.CBOR.Read                  as CBOR
import qualified Data.ByteString                  as BS
import qualified Data.ByteString.Lazy             as BSL
import           System.Exit                      (ExitCode (..), exitSuccess, exitWith)

expectLen :: Int -> Int -> String -> CBOR.Decoder s ()
expectLen actual expected context =
  if actual == expected
     then pure ()
     else fail (context ++ " length " ++ show actual ++ ", expected " ++ show expected)

decodePoint :: CBOR.Decoder s ()
decodePoint = do
  len <- CBOR.decodeListLen
  case len of
    0 -> pure ()
    2 -> do
      _slot <- CBOR.decodeWord64
      hash <- CBOR.decodeBytes
      if BS.length hash == 32
         then pure ()
         else fail ("point hash length " ++ show (BS.length hash) ++ ", expected 32")
    _ -> fail ("point length " ++ show len ++ ", expected 0 or 2")

decodeTip :: CBOR.Decoder s ()
decodeTip = do
  len <- CBOR.decodeListLen
  expectLen len 2 "tip"
  decodePoint
  _height <- CBOR.decodeWord64
  pure ()

decodeHeaderContent :: CBOR.Decoder s ()
decodeHeaderContent = do
  len <- CBOR.decodeListLen
  era <- CBOR.decodeWord8
  case era of
    0 -> do
      expectLen len 2 "byron header_content"
      prefixLen <- CBOR.decodeListLen
      expectLen prefixLen 2 "byron header prefix"
      _prefixA <- CBOR.decodeWord8
      _prefixB <- CBOR.decodeWord64
      _tag <- CBOR.decodeTag
      _bytes <- CBOR.decodeBytes
      pure ()
    _ | era >= 1 && era <= 7 -> do
      expectLen len 2 "header_content"
      _tag <- CBOR.decodeTag
      _bytes <- CBOR.decodeBytes
      pure ()
    _ -> fail ("unknown header_content era variant: " ++ show era)

decoder :: CBOR.Decoder s ()
decoder = do
  len <- CBOR.decodeListLen
  key <- CBOR.decodeWord
  case (len, key) of
    (1, 0) -> pure ()
    (1, 1) -> pure ()
    (3, 2) -> decodeHeaderContent >> decodeTip
    (3, 3) -> decodePoint >> decodeTip
    (2, 4) -> do
      points <- CBOR.decodeListLen
      replicateM_ points decodePoint
    (3, 5) -> decodePoint >> decodeTip
    (2, 6) -> decodeTip
    (1, 7) -> pure ()
    _      -> fail ("unexpected chainsync message (len, key) = (" ++ show len ++ ", " ++ show key ++ ")")

main :: IO ()
main = do
  bytes <- BSL.getContents
  case CBOR.deserialiseFromBytes decoder bytes of
    Left err -> do
      putStrLn ("ERR " ++ show err)
      exitWith (ExitFailure 1)
    Right (rest, ()) | BSL.null rest -> do
      putStrLn "OK"
      exitSuccess
    Right _ -> do
      putStrLn "ERR trailing bytes after message"
      exitWith (ExitFailure 1)

File: dwarf/targets/amaru/src/bin/decode_chainsync.rs
// amaru-mini-protocol-decode-chainsync shim.
// Decodes one ChainSync mini-protocol wire-format message:
//   RequestNext       = [0]
//   AwaitReply        = [1]
//   RollForward       = [2, header_content, tip]
//   RollBackward      = [3, point, tip]
//   FindIntersect     = [4, [point...]]
//   IntersectFound    = [5, point, tip]
//   IntersectNotFound = [6, tip]
//   Done              = [7]

use std::io::{self, Read};
use std::process::ExitCode;

fn expect_len(actual: u64, expected: u64, context: &str) -> Result<(), String> {
    if actual == expected {
        Ok(())
    } else {
        Err(format!("{context} length {actual}, expected {expected}"))
    }
}

fn decode_point(d: &mut minicbor::Decoder<'_>) -> Result<(), String> {
    let len = d
        .array()
        .map_err(|e| format!("point array: {e}"))?
        .ok_or_else(|| "indefinite-length point not allowed".to_string())?;
    match len {
        0 => Ok(()),
        2 => {
            let _slot = d.u64().map_err(|e| format!("point slot: {e}"))?;
            let hash = d.bytes().map_err(|e| format!("point hash: {e}"))?;
            if hash.len() != 32 {
                return Err(format!("point hash length {}, expected 32", hash.len()));
            }
            Ok(())
        }
        _ => Err(format!("point length {len}, expected 0 or 2")),
    }
}

fn decode_tip(d: &mut minicbor::Decoder<'_>) -> Result<(), String> {
    let len = d
        .array()
        .map_err(|e| format!("tip array: {e}"))?
        .ok_or_else(|| "indefinite-length tip not allowed".to_string())?;
    expect_len(len, 2, "tip")?;
    decode_point(d)?;
    let _height = d.u64().map_err(|e| format!("tip block height: {e}"))?;
    Ok(())
}

fn decode_header_content(d: &mut minicbor::Decoder<'_>) -> Result<(), String> {
    let len = d
        .array()
        .map_err(|e| format!("header_content array: {e}"))?
        .ok_or_else(|| "indefinite-length header_content not allowed".to_string())?;
    let era = d.u8().map_err(|e| format!("header_content era: {e}"))?;
    match era {
        0 => {
            expect_len(len, 2, "byron header_content")?;
            let prefix_len = d
                .array()
                .map_err(|e| format!("byron header prefix array: {e}"))?
                .ok_or_else(|| "indefinite-length byron prefix not allowed".to_string())?;
            expect_len(prefix_len, 2, "byron header prefix")?;
            let _prefix_a = d.u8().map_err(|e| format!("byron prefix a: {e}"))?;
            let _prefix_b = d.u64().map_err(|e| format!("byron prefix b: {e}"))?;
            let _tag = d.tag().map_err(|e| format!("byron header tag: {e}"))?;
            let _bytes = d.bytes().map_err(|e| format!("byron header bytes: {e}"))?;
            Ok(())
        }
        1..=7 => {
            expect_len(len, 2, "header_content")?;
            let _tag = d.tag().map_err(|e| format!("header tag: {e}"))?;
            let _bytes = d.bytes().map_err(|e| format!("header bytes: {e}"))?;
            Ok(())
        }
        _ => Err(format!("unknown header_content era variant: {era}")),
    }
}

fn decode(data: &[u8]) -> Result<(), String> {
    let mut d = minicbor::Decoder::new(data);
    let len = d
        .array()
        .map_err(|e| format!("message array: {e}"))?
        .ok_or_else(|| "indefinite-length chainsync message not allowed".to_string())?;
    let key: u32 = d.u32().map_err(|e| format!("message key: {e}"))?;
    match (len, key) {
        (1, 0) | (1, 1) | (1, 7) => {}
        (3, 2) => {
            decode_header_content(&mut d)?;
            decode_tip(&mut d)?;
        }
        (3, 3) | (3, 5) => {
            decode_point(&mut d)?;
            decode_tip(&mut d)?;
        }
        (2, 4) => {
            let points = d
                .array()
                .map_err(|e| format!("find_intersect points array: {e}"))?
                .ok_or_else(|| "indefinite-length point list not allowed".to_string())?;
            for _ in 0..points {
                decode_point(&mut d)?;
            }
        }
        (2, 6) => decode_tip(&mut d)?,
        _ => {
            return Err(format!(
                "unexpected chainsync message (len, key) = ({len}, {key})"
            ))
        }
    }
    if d.position() != data.len() {
        return Err(format!("trailing bytes at position {}", d.position()));
    }
    Ok(())
}

fn main() -> ExitCode {
    let mut buf = Vec::new();
    if let Err(e) = io::stdin().read_to_end(&mut buf) {
        println!("ERR stdin read: {e}");
        return ExitCode::from(1);
    }
    match decode(&buf) {
        Ok(()) => {
            println!("OK");
            ExitCode::from(0)
        }
        Err(msg) => {
            println!("ERR {msg}");
            ExitCode::from(1)
        }
    }
}
